### PR TITLE
DSP: Implement audio codecs (PCM8, PCM16, ADPCM)

### PR DESCRIPTION
**Pull request by MerryMage**
Decoders for the above three codecs, self-explanatory.

### CODE REVIEW

File: src/audio_core/codec.cpp [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+#include "audio_core/codec.h"
+
+#include "common/assert.h"
+#include "common/common_types.h"
+#include "common/math_util.h"
+
+namespace Codec {
+
+StereoBuffer16 DecodeADPCM(const u8 * const data, const size_t sample_count, const std::array<s16, 16>& adpcm_coeff, AdpcmState& state) {
```
Comment (Subv) [inline src/audio_core/codec.cpp]: u8*

File: src/audio_core/codec.cpp [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    // GC-ADPCM with scale factor and variable coefficients.
+    // Frames are 8 bytes long containing 14 samples each.
+    // Samples are 4 bits (one nybble) long.
+
+    constexpr size_t FRAME_LEN = 8;
+    constexpr size_t SAMPLES_PER_FRAME = 14;
+    constexpr std::array<int, 16> SIGNED_NYBBLES {{ 0, 1, 2, 3, 4, 5, 6, 7, -8, -7, -6, -5, -4, -3, -2, -1 }};
```
Comment (Subv) [inline src/audio_core/codec.cpp]: nibbles

File: src/audio_core/codec.cpp (hunk 1 above)
Comment (MerryMage) [inline src/audio_core/codec.cpp]: Oops. Corrected!

File: src/audio_core/codec.h [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+#include "common/common_types.h"
+
+namespace Codec {
+
+/// A variable length buffer of signed PCM16 stereo samples.
+using StereoBuffer16 = std::vector<std::array<s16, 2>>;
+
+/// See: Codec::DecodeADPCM
+struct AdpcmState {
```
Comment (lioncash) [inline src/audio_core/codec.h]: This should likely be upper-cased to ADPCM to stay consistent with the function name `DecodeADPCM`

File: src/audio_core/codec.cpp [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+StereoBuffer16 DecodePCM8(const unsigned num_channels, const u8* const data, const size_t sample_count) {
+    ASSERT(num_channels == 1 || num_channels == 2);
+
+    StereoBuffer16 ret(sample_count);
+
+    if (num_channels == 1) {
+        for (size_t i = 0; i < sample_count; i++) {
+            ret[i][0] = ret[i][1] = SignExtendS8(data[i]);
```
Comment (lioncash) [inline src/audio_core/codec.cpp]:
```cpp
ret[i].fill(SignExtendS8(data[i]))
```

File: src/audio_core/codec.cpp [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+            yn2 = yn1;
+            yn1 = val;
+            return (s16)val;
+        };
+
+        size_t outputi = framei * SAMPLES_PER_FRAME;
+        size_t datai = framei * FRAME_LEN + 1;
+        for (size_t i = 0; i < SAMPLES_PER_FRAME && outputi < sample_count; i += 2) {
+            const s16 sample1 = decode_sample(SIGNED_NYBBLES[data[datai] & 0xF]);
+            ret[outputi][0] = ret[outputi][1] = sample1;
```
Comment (lioncash) [inline src/audio_core/codec.cpp]:
```cpp
ret[outputi].fill(sample1);
```

File: src/audio_core/codec.cpp [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+            outputi++;
+
+            const s16 sample2 = decode_sample(SIGNED_NYBBLES[data[datai] >> 4]);
+            ret[outputi][0] = ret[outputi][1] = sample2;
```
Comment (lioncash) [inline src/audio_core/codec.cpp]:
```cpp
ret[output1].fill(sample2);
```

File: src/audio_core/codec.cpp [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+StereoBuffer16 DecodePCM16(const unsigned num_channels, const u8* const data, const size_t sample_count) {
+    ASSERT(num_channels == 1 || num_channels == 2);
+
+    StereoBuffer16 ret(sample_count);
+
+    if (num_channels == 1) {
+        for (size_t i = 0; i < sample_count; i++) {
+            s16 sample;
+            std::memcpy(&sample, data + i * sizeof(s16), sizeof(s16));
+            ret[i][0] = ret[i][1] = sample;
```
Comment (lioncash) [inline src/audio_core/codec.cpp]:
```cpp
ret[i].fill(sample);
```

**Comment by MerryMage** [comment]
Addressed above comments.

File: src/audio_core/codec.cpp [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+            ret[i][0] = SignExtendS8(data[i * 2 + 0]);
+            ret[i][1] = SignExtendS8(data[i * 2 + 1]);
+        }
+    }
+
+    return ret;
+}
+
+StereoBuffer16 DecodePCM16(const unsigned num_channels, const u8* const data, const size_t sample_count) {
+    ASSERT(num_channels == 1 || num_channels == 2);
```
Comment (Subv) [inline src/audio_core/codec.cpp]: What happens if there are more channels?

Comment (MerryMage) [inline src/audio_core/codec.cpp]: Good question!

Just did a hwtest. The bitfield for determining number of channels is two bits in size, so there are four possibilities (0-3).

* 0, 1, 3: Mono
* 2: Stereo

I expect to handle this before this point (when I'm reading the parameters out of shared memory) so I won't change the code here.

**Comment by Subv** [comment]
LGTM

File: src/audio_core/codec.cpp [hunk 9]
```diff
@@ -0,0 +1,122 @@
+// Copyright 2016 Citra Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#include <array>
+#include <cstring>
+#include <cstddef>
```
Comment (lioncash) [inline src/audio_core/codec.cpp]: cstddef should come before cstring

Comment (MerryMage) [inline src/audio_core/codec.cpp]: amended

**Comment by lioncash** [comment]
LGTM

File: src/audio_core/codec.cpp (hunk 2 above)
Comment (hrydgard) [inline src/audio_core/codec.cpp]: nybbles is kind of logical. bite, byte, nibbles, nybbles :)

**Comment by JamePeng** [comment]
LGTM

**Comment by bunnei** [comment]
LGTM

**Pull request merged**